### PR TITLE
CSpell Action

### PR DESCRIPTION
**Pull request by Skptak**
Splitting up #68 into more manageable single PRs containing just the changes to the individual actions
Will not be doing the YML files one by one, so please reference the #68 actions for proof of working

This exists more or less just to push the cpsell config to users for use with VSCode

Relevant Log can be found [here](https://github.com/Skptak/CI-CD-Github-Actions/actions/runs/5870035221/job/15916259776)

### CODE REVIEW

File: spellings/cspell.config.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+version: '0.2'
+# Allows things like stringLength
+allowCompoundWords: true
+useGitignore: true
+# Could split this up? And do a dictionary for each repo?
+# But feel like if this isn't super slow
+# That having just one single dictionary might be nicer?
+dictionaryDefinitions:
+  - name: freertos-words
+    path: '.cSpellWords.txt'
```
Comment (kstribrnAmzn) [inline spellings/cspell.config.yaml:12]: Should this file be present in this PR? How do I know the dictionary of words hasn't changed?

File: spellings/getFiles [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    # For FreeBSD (default MacOS shell) assume that you have fd
+    # By default only grab the c and h files. Only grab the files with amazon copyright
+    # Put all the files in an array. Uncomment the following line to see the command run.
+    # echo "fd -e c -e h "$excludeDirs" "$excludeFiles" "$includeExtensions" --exec grep -lie \"copyright (.*) 20[0-9]{2} amazon.com\" "
+    # Need this in a set +/- e so it doesn't return an error on files that fail the grep
+    set +e
+    files=$(fd -e c -e h $excludeDirs $excludeFiles $includeExtensions --exec grep -lie "copyright (.*) 20\d\d amazon.com")
+    set -e
+
+# Making the assumption that anybody else running this is a gnu getopt user
```
Comment (kstribrnAmzn) [inline spellings/getFiles:54]: Why can we assume that another user would be using `getopt`?

**Comment by kstribrnAmzn** [comment]
Does this PR force any changes to our existing lexicons setup in our repository? Or is the behavior setup to be identical to the old spell checker and it's dictionaries?

File: spellings/getFiles [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        -ef= | --exclude-files=* )
+            excludeFiles="-E $(echo "${1#*=}" | sed -E 's/,/ -E /g')"
+            shift
+            ;;
+        -ie= | --include-extensions=* )
+            includeExtensions="-e $(echo "${1#*=}" | sed -E 's/,/ -e /g')"
+            shift
+            ;;
+        -h | --help )
+                echo -e "$bashInfo Find all .c and .h files with the Amazon copyright in them $bashEnd"
```
Comment (paulbartell) [inline spellings/getFiles:31]: Nit-pick: Perhaps another scripting language would be more suitable here.

File: spellings/action.yml [hunk 4]
```diff
@@ -2,28 +2,102 @@ name: 'spellings'
...
+          cspell --language-id C --color -c cspell.config.yaml --show-suggestions --show-context "$file"
+          if ! [ $? -eq 0 ]; then
+            exitStatus=1
+          fi
+        done
+        set -e
+
+        echo "::endgroup::"
+        if [ $exitStatus -eq 0 ]; then
+          echo -e "${{ env.bashPass }} ${{ env.stepName }} ${{ env.bashEnd }}"
```
Comment (paulbartell) [inline spellings/action.yml:98]: nit-pick: Perhaps this type of output would be improved if included in a job-summary markdown: https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#adding-a-job-summary

Colorized output isn't very portable, so I would generally prefer not to include it in console logs.

**Review by paulbartell — DISMISSED** [review]

File: spellings/cspell.config.yaml (hunk 1 above)
Comment (Skptak) [reply spellings/cspell.config.yaml:12]: Idea is you would change the dictionary of words if you add in new words, it's the same thing that happens currently with needing to add new words to the lexicon.txt file

File: spellings/getFiles (hunk 2 above)
Comment (Skptak) [reply spellings/getFiles:54]: Well it's an assumption that if you're using command line linux that you are using bash and using the default getopt for it. Where if you're not using it, well then guess this isn't a truly portable script

File: spellings/getFiles (hunk 3 above)
Comment (Skptak) [reply spellings/getFiles:31]: I honestly am debating it, I was sticking to this one for some reason and it's been a huge headache getting it to work for mac and linux and I think the mac implementation is actually buggy...
I was thinking of using golang?

**Comment by Skptak** [comment]
> Does this PR force any changes to our existing lexicons setup in our repository? Or is the behavior setup to be identical to the old spell checker and it's dictionaries?

The idea is to provide a [cspell](https://cspell.org/) configuration that can then be tied into a plugin for [visual studio code](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker) plugin that uses the same format
Where we then use the same dictionary for that spell checker as we do for the rust-spell-check that gets run on PRs as it's faster

File: spellings/action.yml (hunk 4 above)
Comment (Skptak) [reply spellings/action.yml:98]: I had seen issues trying to access things like that from inside of the action, as I think you have to manually pass in your github environment to them to be able to access these?
But think that's an optimization that could be added after the initial merge of this

**Review by A-Zaba — DISMISSED** [review]

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -269,16 +276,58 @@ jobs:
...
     steps:
-      - uses: actions/checkout@v2
-      - uses: actions/checkout@v2
+      - uses: actions/checkout@v3
+
+      - uses: actions/checkout@v3
         with:
-          repository: FreeRTOS/coreMQTT
-          ref: main
+          repository: skptak/coreMQTT
```
Comment (kstribrnAmzn) [inline .github/workflows/test.yml:283]: fix this repo?

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -269,16 +276,58 @@ jobs:
...
           path: coreMQTT
+
+  test-spell-checker-find-mistake:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+
+      - uses: actions/checkout@v3
+        with:
+          repository: skptak/coreMQTT
```
Comment (A-Zaba) [inline .github/workflows/test.yml:299]: Had totally missed this originally, but same issue as Kody's above.

File: .github/workflows/test.yml (hunk 5 above)
Comment (Skptak) [reply .github/workflows/test.yml:283]: Requires a wordlist that doesn't exist. The PR for coreMQTT using this action can be found [here](https://github.com/FreeRTOS/coreMQTT/pull/260)

File: .github/workflows/test.yml (hunk 6 above)
Comment (Skptak) [reply .github/workflows/test.yml:299]: Requires a wordlist and config file that does not exist in current mainline. The PR for coreMQTT using this action can be found [here](https://github.com/FreeRTOS/coreMQTT/pull/260)

File: spellings/getFiles (hunk 3 above)
Comment (Skptak) [reply spellings/getFiles:31]: Ended up changing the way we get files for this, marking it resolved

File: spellings/getFiles (hunk 2 above)
Comment (Skptak) [reply spellings/getFiles:54]: Using fd-find now, removed this script entirely.

**Review by A-Zaba — DISMISSED** [review]

**Review by A-Zaba — APPROVED** [review]

**Review by rawalexe — APPROVED** [review]

**Pull request merged**